lib/src/source/hosted.dart

```diff
@@ -354,11 +363,13 @@ class BoundHostedSource extends CachedSource {
 
   /// Downloads package [package] at [version] from [server], and unpacks it
   /// into [destPath].
-  Future _download(
-      String server, String package, Version version, String destPath) async {
-    var url = Uri.parse('$server/packages/$package/versions/$version.tar.gz');
+  Future _download(PackageId id, String destPath) async {
+    final versions = await _scheduler.schedule(id.toRef());
+    final url = versions[id]?.archiveUrl ??
+        (throw PackageNotFoundException(
```

Comment:
Being unable to find the version and being unable to find the archiveUrl is two different issues.

If we want find the package we should fallback to `PackageNotFoundException`, if we can't find the `archiveUrl` I propose that we could fallback to the old URL pattern?

Honestly, I would prefer to avoid this, but there might be 3rd party pub servers out there which doesn't have the `archiveUrl`.

Reply:
Ok
Done.

Now it will fall back to the old URL pattern if the version exists in the version info from the api/package/versions but has no archive url. If it does not exist in the version listing we throw up front.

You like?